lib/failure_detector.ml

```diff
@@ -0,0 +1,138 @@
+open Base.Hashtbl
+
+type message =
+  | Ping
+  | Acknowledge
+  | PingRequest of Address.t
+[@@deriving bin_io]
+
+type config = {
+  protocol_period : int;
+  round_trip_time : int;
+  timeout : int;
+  mutable peers_to_ping : int;
+}
+
+type t = {
+  config : config;
+  (* TODO: I am totally not sure about this type *)
+  acknowledges : (int, unit Lwt_condition.t) Base.Hashtbl.t;
+  mutable peers : Peer.t list;
+  mutable sequence_number : int;
+}
+
+(* Helper to increase the round *)
+let next_seq_no t =
+  let sequence_number = t.sequence_number in
+  t.sequence_number <- t.sequence_number + 1;
+  sequence_number
+
+let send_message client (recipient : Peer.t) message =
+  let message = Util.Encoding.pack bin_writer_message message in
+  Client.send_to client message recipient
+
+let send_ping_to client recipient = send_message client recipient Ping
+
+let send_acknowledge_to client recipient =
+  send_message client recipient Acknowledge
+
+let send_ping_request_to client recipient =
+  send_message client recipient (PingRequest recipient.socket_address)
+
+(* TODO: Totally not sure about this...
+   I don't know if I can get rid of it and only use `wait_ack_timeout` instead
+    But I feel like Lwt_condition allows ms to manage async thread and could be useful *)
+let[@warning "-32"] [@warning "-39"] rec wait_ack t sequence_number =
+  let cond =
+    find_or_add t.acknowledges sequence_number ~default:Lwt_condition.create
+  in
+  Lwt_condition.wait cond
+
+(* TODO: An async thread should be started here, with a basic sleep, but not sure at all...
+     And we should return Either Ok Timeout *)
+let wait_ack_timeout t sequence_number timeout =
+  Lwt.pick
+    [
+      (let _ = Lwt_unix.sleep (Float.of_int timeout) in
+       Lwt.return @@ Result.Error "timeout");
+      (let _ = wait_ack t sequence_number in
+       Lwt.return @@ Result.ok "received ACK");
+    ]
+
+(* Basic Knuth shuffle => https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle *)
+let knuth_shuffle known_peers =
+  let initial_array = Array.length (Array.of_list known_peers) in
+  let shuffled_array = Array.copy (Array.of_list known_peers) in
```

Comment:
It's probably "expensive" to convert the same list to an array twice, you probably want to use `Array.copy initial_array` instead.

Reply:
Nice catch! 👍 
changed to:
```ocaml
  let shuffled_array = Array.copy (Array.of_list known_peers) in
  let initial_array_length = Array.length shuffled_array in
```
to avoid duplicate `of_list`